src/xcode/ENA/ENA/Source/Scenes/TicketValidation/CertificateSelection/TicketValidationCertificateSelectionViewModel.swift

```diff
@@ -0,0 +1,218 @@
+				supportedCertificateTypes.append(AppStrings.TicketValidation.SupportedCertificateType.pcrTestCertificate)
+			}
+			// if type contains tr, all RAT tests shall pass the filter
+			if certificateTypes.contains("tr") {
+				supportedHealthCertificates.append(contentsOf: healthCertifiedPersonCertificates.filter { $0.testEntry != nil && $0.testEntry?.typeOfTest == TestEntry.antigenTypeString })
+				supportedCertificateTypes.append(AppStrings.TicketValidation.SupportedCertificateType.ratTestCertificate)
+			}
+		} else {
+			// if type is nil or empty, then there is no filtering by type
+			supportedHealthCertificates = healthCertifiedPersonCertificates
+		}
+		
+		// sorting on the basis of certificate type
+		supportedHealthCertificates = supportedHealthCertificates.sorted(by: >)
+		
+		// creating service provider requirements description
+		let serviceProviderRequirementsDescription = generateServiceProviderRequirementsString(supportedCertificateTypes: supportedCertificateTypes, validationConditions: validationConditions)
+		
+		// finding health certified person
+		var healthCertifiedPerson: HealthCertifiedPerson?
+		self.healthCertificateService.healthCertifiedPersons.forEach { certifiedPerson in
+			supportedHealthCertificates.forEach { healthCertificate in
+				if certifiedPerson.healthCertificates.contains(healthCertificate) {
+					healthCertifiedPerson = certifiedPerson
+				}
+			}
+		}
+		
+		// setting up view model
+		if supportedHealthCertificates.isEmpty {
+			dynamicTableViewModel = dynamicTableViewModelNoSupportedCertificate(serviceProviderRequirementsDescription: serviceProviderRequirementsDescription)
+		} else {
+			dynamicTableViewModel = dynamicTableViewModelSupportedHealthCertificates(healthCertifiedPerson: healthCertifiedPerson, serviceProviderRequirementsDescription: serviceProviderRequirementsDescription)
+		}
+	}
+		
+	private func generateServiceProviderRequirementsString(supportedCertificateTypes: [String], validationConditions: ValidationConditions) -> String {
+		var serviceProviderRequirementsDescription: String = ""
+
+		serviceProviderRequirementsDescription += supportedCertificateTypes.joined(separator: ", ")
+		if let dateOfBirth = validationConditions.dob {
+			serviceProviderRequirementsDescription += String(format: AppStrings.TicketValidation.CertificateSelection.dateOfBirth, dateOfBirth)
+		}
+		if let familyName = validationConditions.fnt, let givenName = validationConditions.gnt {
+			serviceProviderRequirementsDescription += "\n\(familyName)<<\(givenName)"
+		} else if let familyName = validationConditions.fnt {
+			serviceProviderRequirementsDescription += "\n\(familyName)<<"
+		} else if let givenName = validationConditions.gnt {
+			serviceProviderRequirementsDescription += "\n<<\(givenName)"
+		}
+		
+		return serviceProviderRequirementsDescription
+	}
```

Comment:
I think we should have unit tests for these two funcs. Maybe you could even extract the to a separate struct or class to make them testable.